mlir/lib/Dialect/Rock/utility/transformMapUtils.cpp

```diff
@@ -346,9 +348,249 @@ propagateUnmergeVectorization(T &&dimAndLength,
   return result;
 }
 
+// Key data structures for vectorization analysis
+using MergePairKey = std::pair<TransformMapAttr, TransformAttr>;
+
+using DimToMergeMap =
+    llvm::DenseMap<uint32_t, std::pair<TransformMapAttr, TransformAttr>>;
+
+using DimRemapMap = llvm::DenseMap<uint32_t, uint32_t>;
+
+using ContiguousMergesMap =
+    llvm::DenseMap<MergePairKey, SmallVector<SmallVector<uint32_t>>>;
+
+// Utility data structure to save information about a specific merge
+struct MergeInfo {
+  MergeInfo() {}
+  MergeInfo(MergePairKey mergePair) {
+    auto mergeLowerDims = mergePair.second.getLowerDims();
+    for (auto pair : llvm::zip(mergeLowerDims, mergePair.second.getParams())) {
+      paramMap[std::get<0>(pair)] = std::get<1>(pair);
+    }
+
+    for (size_t j = 0; j < mergeLowerDims.size(); j++) {
+      dimPosition[mergeLowerDims[j]] = j;
+    }
+  }
+
+  // Given a dimension this map tells us what is the parameter
+  // associated to that dimension
+  llvm::DenseMap<uint32_t, int64_t> paramMap;
+  // Given a dimension, this map tells us what is its position
+  // in the dimension vector (e.g., if mergeDims = [0,2,3], dimPosition(2)=1)
+  llvm::DenseMap<uint32_t, int64_t> dimPosition;
+};
+
+// Given dimensions and parameters of an unmerge, scan the merges
+// to flag group of dimensions that are known to be contiguous
+// in the unmerge
```

Comment:
I'm broadly suspicious of how intertwined and stateful this code is. Would it be possible to have a function that takes a single `TransformMapAttr` and some sort of `ContiguousMergesInfo` and then returns a _new_ `ContiguousMergesInfo` that reflects the effects of the transformations?